Use Mesh instead of index

nilsvu opened the pull request:
## Proposed changes

This PR summarises #543, #659 and #711, and refactors the code base to use the `Mesh` type and the `Spectral` namespace.

### `Mesh` type

- #659 introduced the `Mesh<Dim>` type to carry information on the computational grid. In particular, this type provides everything that is required to interpret `DataVector`s on the grid, i.e. the choice of spectral basis functions and quadrature in addition to the number of grid points in each dimension.

### `Spectral` namespace

- #543 introduced the `Spectral` namespace with low-level functions to compute collocation points, quadrature weights and such. There is a version of these functions templated on the spectral basis and quadrature method, and another version that takes a `Mesh<1>`.
- Available basis functions are for the moment only Legendre polynomials (`Spectral::Basis::Legendre`), with the option to add, for instance, Chebyshev polynomials (#716) or spherical harmonics later.
- Both Gauss-Lobatto quadrature and Gauss quadrature are available in `Spectral::Quadrature`.

### What breaks:

- If you are working with an `Index` you'll probably want to use `Mesh` instead wherever it represents the computational grid. In cases where a function really only needs the extents of the grid you can still work with `Index`. You'll find documentation here: https://spectre-code.org/classmesh
- Replace `Tags::Extents` with `Tags::Mesh` in the DataBox. From the stored `Mesh` you can get the `Index<Dim> Mesh::extents()` if necessary.
- The `Basis::lgl` namespace was superseded by `Spectral`, so these function calls must be replaced. For instance to retrieve a differentiation matrix in dimension `d` of the grid call `Spectral::differentiation_matrix(mesh.slice_through(d))`.

### Types of changes:

- [x] New feature

### Component:

- [x] Code

### Code review checklist

- [x] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For instructions on how to perform the CI checks locally refer to the [Dev guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [x] The code is documented and the documentation renders correctly. Run `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`. Then open `index.html`.
- [x] The code follows the stylistic and code quality guidelines listed in the [code review guide](https://spectre-code.org/code_review_guide.html).

nilsvu commented:
Add the _work in progress_ label please.

teukolsky commented on src/NumericalAlgorithms/Spectral/Spectral.tpp:
```diff
@@ -0,0 +1,259 @@
+      const double p_n =
+          ((2 * k - 1) * x[j] * p_n_minus1 - (k - 1) * p_n_minus2) / k;
+      const double gamma = (k == number_of_pts - 1)
+                               ? 2.0 / (number_of_pts - 1.0)
```
I think the special case when gamma = 2.0/(number_of_pts-1.0) is only for GLL, not GL

nilsvu commented:
Oh I see we also have a label _dependent_, please also add that (depends on #631).

mar-celine commented on src/DataStructures/Mesh.hpp:
```diff
@@ -0,0 +1,34 @@
+  template <typename... I, Requires<(sizeof...(I) > 1)> = nullptr>
+  explicit Mesh(I... i) : Index<dim>(i...) {}
+  explicit Mesh(std::array<size_t, dim> i) : Index<dim>(std::move(i)) {}
+};
```
this will also need a PUP function

nilsvu replied:
Implemented one, hope I got it about right ^^

mar-celine commented on src/DataStructures/Mesh.hpp:
```diff
@@ -0,0 +1,34 @@
+
+template <class... Bases>
+bool operator==(const Mesh<Bases...>& lhs, const Mesh<Bases...>& rhs) noexcept {
+  return std::equal(lhs.begin(), lhs.end(), rhs.begin());
```
use the == operator for Index

mar-celine commented on src/NumericalAlgorithms/Spectral/Spectral.hpp:
```diff
@@ -0,0 +1,119 @@
+
+#include "DataStructures/DataVector.hpp"
+
+class Matrix;
```
cond this

nilsvu replied:
Sorry what do you mean?

osheamonn replied:
Doxygen can sometimes mess up when presented with forward declarations so we do this to ignore them in the docs:
```
/// \cond
class Matrix; 
/// \endcond 
```

mar-celine commented on src/NumericalAlgorithms/Spectral/Spectral.hpp:
```diff
@@ -0,0 +1,119 @@
+
+namespace detail {
+template <typename BasisType>
+class CollocationPointsAndWeights {
```
can be moved to a cpp file

nilsvu replied:
I'm not quite sure but I think not, since we want to specialise it in basis-specific cpp files such as `Legendre.cpp` (to provide a specialised constructor). I could move it to `Spectral.tpp` I guess, which is included in each of these basis-specific files anyway.

mar-celine commented on src/NumericalAlgorithms/Spectral/Spectral.tpp:
```diff
@@ -0,0 +1,259 @@
+// Distributed under the MIT License.
```
lets talk about this file monday at 2pm eastern

nilsvu commented:
The `logical_coordinates` function in `Domain/LogicalCoordinates.cpp` makes use of the new `Mesh` type now, with runtime switches between the bases and quadratures. There is also a new `Tags::Mesh` data box tag that is used by the `Tags::LogicalCoordinates` compute item.

Could you have a look over this (in particular `Mesh.hpp` and `LogicalCoordinates.cpp`) and let me know if this is sort of what you had in mind, what could be done better perhaps, and whether I should proceed along these lines to refactor more code (derivatives, integrals and such)?

**Edit:**

- The runtime switches are now actually performed in `Mesh.cpp`, which I think could be the way to go: Now, `Mesh` mediates between the low-level `Spectral` namespace and higher-level functions that compute derivatives and such in multiple dimensions (see updated PR description above). It knows about all dimensions of the grid and can therefore stitch together also information from multi-dimensional basis functions (like spherical harmonics) should this become necessary. And it does this at runtime to handle also the mesh on mortars, where the choice of quadrature is dynamic.
- You can see the `Mesh` in action in e.g. `logical_coordinates`, that just collects the collocation points in each dimension, or in `definite_integral`.

nilsvu commented:
Btw. this PR is touching a lot of files, so it will be a pain to rebase if we keep it around for too long. Would be great if we could move it along sooner rather than later therefore 😄

wthrowe commented:
You should not have a Spectral.tpp file.  Anything basis-agnostic should be in a Spectral.cpp with explicit instantiations and anything basis-specific should be in a file for that basis.

nilsvu commented:
Edited my comment above, this PR should be ready for review now.

@wthrowe About your comment on `Spectral.tpp`: It implements algorithms that proceed the same for each basis. These functions are templated on `Spectral::Basis` and `Spectral::Quadrature`, so my intention was to include them in each basis-specific `.cpp` file, such as `Legendre.cpp`, to generate these functions for that basis. I think this is actually the correct use case for `.tpp` files, or would you suggest to implement this differently?

kidder commented on src/DataStructures/Mesh.hpp:
```diff
@@ -0,0 +1,87 @@
+    quadratures_.fill(uniform_quadrature);
+  }
+
+  explicit Mesh(std::array<size_t, Dim> extents,
```
no explicit

nilsvu replied:
could you explain a bit the reasoning where we want this and where we don't?

nilsdeppe replied:
http://en.cppreference.com/w/cpp/language/explicit

we don't want converting constructors in general

kidder commented on src/DataStructures/Mesh.hpp:
```diff
@@ -0,0 +1,87 @@
+class Mesh : public Index<Dim> {
+ public:
+  explicit Mesh(
+      const size_t uniform_extents = std::numeric_limits<size_t>::max(),
```
remove uniform_ from the input names and remove the default arguments

nilsvu replied:
I put the defaults here since LGL is actually the default mesh used everywhere in the code. So when a `Mesh` is required where people don't really care about the choice of basis, LGL is a reasonable choice. Moste (all?) functionality should work for any basis anyway, so it's OK if people don't care about this too much until it becomes necessary. So e.g. in tests where other functionality is tested, such as making sure some compute items work, this should be independent of the choice of basis and the defaults can be used. But then to make sure some functionality actually works on mixed meshes, this is where the choice of bases is specified explicitly.

I also set it up this way to further decouple the low-level `Spectral` namespace from code that uses the `Mesh`: Usually code only needs to work with the `Mesh` and doesn't even need to include `Spectral.hpp`. Only when it becomes necessary to work with a specific basis and such does it have to make use of the `Spectral` namespace.

Do you agree with this architecture, or have another opinion? We can discuss in the hangout, too.

nilsdeppe replied:
We want to be able to forward declare the enum class, so that DataStructures isn't tied to NumericalAlgorithms too tightly

nilsvu replied:
Hmm I see, good point. Since the `Mesh` is all about placing the `Spectral` algorithms into context, we could move it from `DataStructures` to `Spectral`. Then we could keep the defaults and also the member functions. Or is there another way to keep the defaults?

nilsdeppe replied:
I've thought about that a little bit, but I'm not sure I see having to specify the basis and quadrature as a bad thing, actually. It adds code but also removes any assumptions

Edit:
Specifically, since this would result in dispatching to the wrong differentiation matrix it would lead to a very confusing and difficult to track down bug.

nilsvu replied:
Actually objects such as the differentiation matrix are obtained from the mesh that is passed alongside every `DataVector`, so it should always be correct (A `DataVector` that represents field values has no meaning without a `Mesh` anyway, unless the programmer implicitly assumes an LGL mesh, which they should not). But I agree that what you proposed is better, refactoring now.

Only one issue: If I am not mistaken we need a default constructor for `Mesh` that takes no arguments, so how can we set this up?

nilsdeppe replied:
Okay, great :)

I tested this yesterday to make sure everything would work before we suggested the change. You can just use `Mesh() noexcept = default;` and then do `std::array<Basis, Dim> basis_{};`

edit:
You do not need the full definition of an enum to hold it in an array, since the size is know just from the forward declaration

kidder commented on src/DataStructures/Mesh.hpp:
```diff
@@ -0,0 +1,87 @@
+      Spectral::Basis uniform_basis = Spectral::Basis::Legendre,
+      Spectral::Quadrature uniform_quadrature =
+          Spectral::Quadrature::GaussLobatto)
+      : Index<Dim>(std::move(uniform_extents)) {
```
noexcept everything

kidder commented on src/DataStructures/Mesh.hpp:
```diff
@@ -0,0 +1,87 @@
+        bases_(std::move(bases)),
+        quadratures_(std::move(quadratures)) {}
+
+  Spectral::Basis basis(size_t d) const noexcept { return bases_[d]; }
```
just provide accessors for the arrays

kidder commented on src/DataStructures/Mesh.hpp:
```diff
@@ -0,0 +1,87 @@
+  /// \param d the dimension to remove.
+  template <size_t N = Dim, Requires<(N > 0)> = nullptr>
+  Mesh<Dim - 1> slice_away(const size_t d) const noexcept {
+    const auto slice_extents = Index<Dim>::slice_away(d).indices();
```
move function definition to Mesh.cpp

kidder commented on src/DataStructures/Mesh.cpp:
```diff
@@ -0,0 +1,134 @@
+}
+
+template <size_t Dim>
+const Matrix& Mesh<Dim>::differentiation_matrix(const size_t d) const noexcept {
```
@nilsdeppe will provide a way to combine these functions

kidder commented on src/DataStructures/Mesh.cpp:
```diff
@@ -0,0 +1,134 @@
+template <size_t N>
+bool operator==(const Mesh<N>& lhs, const Mesh<N>& rhs) noexcept {
+  return lhs.bases_ == rhs.bases_ and lhs.quadratures_ == rhs.quadratures_ and
+         static_cast<Index<N>>(lhs) == static_cast<Index<N>>(rhs);
```
`static_cast<const Index<N>&>>(lhs) ...`

kidder commented on src/DataStructures/Mesh.hpp:
```diff
@@ -0,0 +1,87 @@
+  friend bool operator==(const Mesh<N>& lhs, const Mesh<N>& rhs) noexcept;
+
+  template <size_t N>
+  friend bool operator!=(const Mesh<N>& lhs, const Mesh<N>& rhs) noexcept;
```
doesn't need to be a friend

kidder commented on src/DataStructures/Mesh.hpp:
```diff
@@ -0,0 +1,87 @@
+
+  /// \cond HIDDEN_SYMBOLS
+  template <size_t N>
+  friend bool operator==(const Mesh<N>& lhs, const Mesh<N>& rhs) noexcept;
```
after changing the accessors, this doesn't need to be a friend

kidder commented on src/DataStructures/Mesh.hpp:
```diff
@@ -0,0 +1,87 @@
+  const Matrix& differentiation_matrix(const size_t d) const noexcept;
+  const Matrix& linear_filter_matrix(const size_t d) const noexcept;
+
+  /// \cond
```
don't cond the pup function

kidder commented on src/DataStructures/Mesh.hpp:
```diff
@@ -0,0 +1,87 @@
+    return Mesh<Dim - 1>(slice_extents, slice_bases, slice_quadratures);
+  }
+
+  const DataVector& collocation_points(const size_t d) const noexcept;
```
these should be free functions taking a Mesh, and in Spectral.hpp

kidder commented on src/DataStructures/Mesh.hpp:
```diff
@@ -0,0 +1,87 @@
+#pragma once
+
+#include "DataStructures/Index.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
```
don't include, just forward declare the enums

kidder commented on src/DataStructures/Mesh.cpp:
```diff
@@ -0,0 +1,134 @@
+  GEN_OP(==, DIM(data))           \
+  GEN_OP(!=, DIM(data))
+
+GENERATE_INSTANTIATIONS(INSTANTIATE, (0, 1, 2, 3, 4))
```
just up to 3

kidder reviewed (changes requested):
Create a PR with just Mesh.hpp Mesh.cpp Test_Mesh.cpp and a simple Spectral.hpp that contains only the enums

Thank you.  Also do you want to meet after 3PM ET Thursday or  at 10 AM ET Friday, or sometime Monday?

nilsvu commented:
@kidder let's do 3pm ET today :)

mar-celine commented on src/DataStructures/Mesh.hpp:
```diff
@@ -9,14 +9,17 @@
 #include "DataStructures/Index.hpp"
 #include "NumericalAlgorithms/Spectral/Spectral.hpp"
 
+struct Matrix;
```
remember to cond

mar-celine commented on src/DataStructures/Mesh.hpp:
```diff
@@ -7,28 +7,26 @@
-          Spectral::Quadrature::GaussLobatto) noexcept
+  Mesh() noexcept = default;
+
+  explicit Mesh(const size_t uniform_extents, Spectral::Basis uniform_basis,
```
no need for explicit here

mar-celine commented on src/DataStructures/Mesh.hpp:
```diff
@@ -7,28 +7,26 @@
-                Spectral::Basis uniform_basis = Spectral::Basis::Legendre,
-                Spectral::Quadrature uniform_quadrature =
-                    Spectral::Quadrature::GaussLobatto) noexcept
+  explicit Mesh(std::array<size_t, Dim> extents, Spectral::Basis uniform_basis,
```
similarly

mar-celine commented on src/NumericalAlgorithms/Spectral/Spectral.hpp:
```diff
@@ -88,6 +100,10 @@ const Matrix& spectral_to_grid_points_matrix(size_t number_of_pts) noexcept;
 const Matrix& linear_filter_matrix(size_t number_of_pts) noexcept;
 
+template <size_t Dim>
+const Matrix& linear_filter_matrix(const Mesh<Dim>& mesh,
```
look into tests for grid to spectral and spectral to grid functions

mar-celine reviewed (commented):
rebase after newton-raphson one is merged

nilsvu commented:
Rewrote the history. Will wait for #631 now before rebasing on `develop`. Then, I'll open separate PRs to merge this piece by piece. The first two `Mesh` and `Spectral` commits are independent of each other, the third implements the `Spectral` functions using `Mesh`, and the fourth refactors the code base to use both.

nilsvu committed:
Remove Tags::Extents and replace with Tags::Mesh

In addition to the extents of the grid in each dimension, Mesh carries
the choice of basis functions and quadrature. All functions that need
Spectral quantities work with Mesh now, and where you need only the
extents you can still get mesh.extents().

nilsvu commented:
Should be ready, fixing clang-tidy and iwyu tomorrow. As mentioned in the call today, I'd appreciate if we could review+merge this quickly, since it touches a _lot_ of files and will be a pain to rebase if left idle for too long.

wthrowe commented:
Question about porting to the new system: Should functions that previously took extents as an Index and did not make any assumptions about the basis continue to take an Index or switch to a Mesh?

nilsvu commented:
I think a function that will only ever depend on the extents of the grid can continue to take an Index, e.g. your `apply_matrices` is a candidate.

I find it difficult to estimate where additional logic will be necessary for higher-dimensional basis functions, i.e. where an Index is not sufficient. To make this clearer I might implement spherical harmonics at some point, but this has quite low priority of course.

Generally when in doubt, and when it's really the computational grid or a slice through it that's represented, it's probably easiest to work with a `Mesh`. The latter also has a more suitable interface, e.g. `Mesh::number_of_grid_points()` instead of `Index::product()`.

nilsvu commented:
Should I move `Mesh.hpp` from `DataStructures/` to `Domain/` (in a separate PR)? Could be useful to have utility functions such as an overload `Mesh<Dim>::extents(const Direction<Dim>&)` that depend on `Domain` types, and we probably don't want to link `Domain` into `DataStructures`.

nilsdeppe commented:
I agree with not linking DataStructures to Domain. I think the only thing DataStructures should depend on is `Utilities`. @wthrowe your thoughts?

kidder commented:
I agree about moving Mesh to Domain

kidder commented on src/DataStructures/DataVector.cpp:
```diff
@@ -7,13 +7,21 @@
       data_(owned_data_.data(), size_) {}
 
+template <size_t Dim>
+DataVector::DataVector(const Mesh<Dim>& mesh, const double value)
```
of course this constructor will make DataVector depend upon Mesh...
thoughts @nilsdeppe ?

nilsdeppe replied:
I don't see any reason to have this constructor, same with the one in `Variables`, especially with moving `Mesh` to `Domain`. Also as you pointed out in another PR, `DataVector` doesn't need to be at points represented by a `Mesh`.

nilsvu replied:
Without this constructor there will be many calls to `Mesh::number_of_grid_points()` throughout the code. How about I rename that to `Mesh::num_points()` to make it less verbose? `num` is a common abbreviation already, the `of` is implied and the `grid` is redundant with calling it on a `mesh`.

kidder commented on src/DataStructures/Variables.hpp:
```diff
@@ -110,10 +111,19 @@ class Variables<tmpl::list<Tags...>> {
 
+  /// Uniformly initialize all tensors with the Mesh::number_of_grid_points() of
+  /// the given \p mesh and the \p value.
+  template <size_t Dim>
```
as will this constructor...

kidder commented on src/Domain/LogicalCoordinates.hpp:
```diff
@@ -42,5 +45,5 @@ tnsr::I<DataVector, VolumeDim, Frame::Logical> logical_coordinates(
  */
 template <size_t VolumeDim>
 tnsr::I<DataVector, VolumeDim, Frame::Logical> interface_logical_coordinates(
```
should this be renamed face_logical_coordinates?  this terminology may be a relic from the Bitbucket where an Interface was a class.  Should we reserve interface for internal faces, boundary_face for external faces, and face for a generic face?  @wthrowe @nilsdeppe thoughts?

nilsdeppe replied:
I like having three different, clearly defined terms to differentiate between the types of interfaces. Your suggestion sounds good to me :)

wthrowe replied:
We need better names for these things, and these suggestions are fine with me.  (The communication code is very inconsistent with all these terms so that isn't going to be very useful for inspiration here.)

nilsvu replied:
To refine the wording, how about _boundary_ (omit the _face_ here)

wthrowe commented:
I agree with @nilsdeppe that we should keep Domain stuff out of DataStructures.  Domain currently is only used in `orient_variables_on_slice`.  I'm not sure how to deal with that case, but let's not add any more.

nilsvu commented:
@wthrowe you could move VariableHelpers.hpp to Domain, or a subset of it

nilsvu commented:
Removed the DataVector and Variables constructors and squashed. Could we meet in a hangout if there's anything else you'd like to change so that we can get this in quickly? Changes to Mesh will be in a separate PR if we decide to move it to Domain.

kidder commented on src/Domain/LogicalCoordinates.cpp:
```diff
@@ -2,57 +2,66 @@
   return logical_x;
 }
 
+template <>
```
why is this here?

nilsdeppe replied:
If this compiles then I'll be confused :P

nilsvu replied:
Are you asking why I added the 1D specialization? `Mesh<Dim>::slice_through()` is available only for `Dim>0`.

And yes, it compiles fine.

nilsdeppe replied:
Ah, it's in the cpp file. I was wondering how it new about the function. I'd find it clearer to have it below the general implementation personally. It might be useful to add a short comment saying the mesh slice through thing.

nilsvu replied:
👍 I'll do that

kidder reviewed (changes requested):
squash

kidder commented:
@nilsdeppe please look at

nilsdeppe commented on src/Domain/LogicalCoordinates.hpp:
```diff
@@ -29,11 +32,11 @@ class Direction;
 /*!
  * \ingroup ComputationalDomainGroup
- * \brief Compute the logical coordinates on a face of an Element
+ * \brief Computes the logical coordinates on a face of an Element.
```
Computes -> Compute
Functions are "active"

nilsdeppe commented on src/Domain/LogicalCoordinates.hpp:
```diff
@@ -11,16 +11,19 @@
 /*!
  * \ingroup ComputationalDomainGroup
- * \brief Compute the Legendre-Gauss-Lobatto coordinates in an Element
+ * \brief Computes the logical coordinates in an Element.
```
Computes -> Compute

Apply liberally throughout your PR

nilsdeppe commented on src/Domain/Tags.hpp:
```diff
@@ -408,8 +434,8 @@ struct Subitems<
       auto sub_var_it = sub_var.begin();
-      for (auto vars_it = parent_vars.begin();
-           vars_it != parent_vars.end(); ++vars_it, ++sub_var_it) {
+      for (auto vars_it = parent_vars.begin(); vars_it != parent_vars.end();
```
Why?

nilsdeppe commented on src/Domain/Tags.hpp:
```diff
@@ -362,13 +374,27 @@ struct InterfaceExtents : db::ComputeTag {
   using argument_tags = tmpl::list<Direction<VolumeDim>, Extents<VolumeDim>>;
   using volume_tags = tmpl::list<Extents<VolumeDim>>;
 };
+template <size_t VolumeDim>
```
@wthrowe is there somewhere where interface tags are being documented and this should be added?

wthrowe replied:
We don't currently have such a place, no.  Most of them just do the same thing as they do in the volume, so I don't think anything is needed for those, but we should document the special ones somewhere.  Adding it to the documentation for the wrapped tag (so `Tags::Mesh` here) seems reasonable to me.  I think that's where I'd be most likely to look.

nilsdeppe replied:
Okay, thanks for clarifying that :)

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/InitializeElement.hpp:
```diff
@@ -203,6 +207,9 @@ struct InitializeElement {
       }
     }
 
+    Variables<db::wrap_tags_in<Tags::dt, variables_tags>> dt_variables{
```
Don't do this. Why? This change has nothing to do with this commit or this PR

nilsdeppe commented on tests/Unit/DataStructures/Test_OrientVariablesOnSlice.cpp:
```diff
@@ -325,29 +344,47 @@ void test_3d_rotated(const Map& my_map, const std::array<size_t, 8>& my_corners,
-    ext.emplace_back(center_x_extents, upper_y_extents, center_z_extents);
-    ext.emplace_back(center_x_extents, center_y_extents, upper_z_extents);
-    return ext;
+    std::vector<Mesh<3>> mshs;
```
`local_meshes`

nilsdeppe commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunication.cpp:
```diff
@@ -259,14 +261,15 @@ SPECTRE_TEST_CASE("Unit.DiscontinuousGalerkin.Actions.FluxCommunication",
                        std::move(mortar_history),
                        std::move(mortar_next_temporal_ids));
-  }();
+  }
```
Why? No...

nilsdeppe commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunication.cpp:
```diff
@@ -312,14 +315,15 @@ SPECTRE_TEST_CASE("Unit.DiscontinuousGalerkin.Actions.FluxCommunication",
-                       std::move(normal_dot_fluxes_map),
-                       std::move(other_data_map), std::move(mortar_history));
+    auto box =
+        db::create<db::AddSimpleTags<
```
Nope, why did so much change?

nilsdeppe commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunication.cpp:
```diff
@@ -351,79 +355,83 @@ SPECTRE_TEST_CASE("Unit.DiscontinuousGalerkin.Actions.FluxCommunication",
           1);
   };
 
-  db::mutate<mortar_data_tag>(make_not_null(&received_box), [
```
All this fro index -> mesh? I doubt that... Fix it.

nilsdeppe commented on src/Evolution/Systems/CurvedScalarWave/Equations.cpp:
```diff
@@ -91,10 +91,9 @@ using derivative_frame = Frame::Inertial;
-      const Index<DIM(data)>& extents,                                       \
+      const Variables<variables_tags<DIM(data)>>&, const Mesh<DIM(data)>&,   \
       const InverseJacobian<DataVector, DIM(data), Frame::Logical,           \
-                            derivative_frame>& inverse_jacobian) noexcept;
```
Again? Please stop mucking around with code for no reason. It makes reviews of things that are basically renames way too difficult.

nilsdeppe commented on src/NumericalAlgorithms/LinearOperators/PartialDerivatives.tpp:
```diff
@@ -23,21 +23,22 @@ struct LogicalImpl;
-  return partial_derivatives_detail::LogicalImpl<
-      Dim, VariableTags, DerivativeTags>::apply(u, extents);
+    const Variables<VariableTags>& u, const Mesh<Dim>& mesh) noexcept {
+  return partial_derivatives_detail::LogicalImpl<Dim, VariableTags,
```
Make this change smaller

nilsdeppe commented on src/NumericalAlgorithms/LinearOperators/PartialDerivatives.tpp:
```diff
@@ -128,48 +130,48 @@ template <typename VariableTags, typename DerivativeTags>
         u.number_of_grid_points());
     const Matrix& differentiation_matrix_eta =
-        Basis::lgl::differentiation_matrix(extents[1]);
+        Spectral::differentiation_matrix(mesh.slice_through(1));
```
looking at this, I find the change really confusing. Would `get_submesh` have been a better term than `slice_through`?

nilsvu replied:
The `Spectral`+`Mesh` interface we discussed in #711 and agreed on this. If in actual use now this turns out to be unclear let’s rethink it in a separate issue or PR if it’s important enough.

nilsdeppe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp:
```diff
@@ -91,16 +91,16 @@ struct ReceiveDataForFluxes {
               tuples::get<typename flux_comm_types::FluxesTag>(inboxes);
           for (auto received_data = inbox.begin();
                received_data != inbox.end() and
-                   received_data->first < local_next_temporal_id;
```
Why?

nilsdeppe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp:
```diff
@@ -91,16 +91,16 @@ struct ReceiveDataForFluxes {
                      "Expected data at "
-                     << neighbor_next_temporal_ids->at(mortar_id)
-                     << " but received at " << receive_temporal_id);
+                         << neighbor_next_temporal_ids->at(mortar_id)
```
why?

nilsdeppe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp:
```diff
@@ -126,8 +126,9 @@ struct ReceiveDataForFluxes {
               "Shouldn't have received data that depended upon the step being "
-              "taken: Received data at " << inbox.begin()->first
-              << " while stepping to " << local_next_temporal_id);
+              "taken: Received data at "
```
why?

nilsdeppe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp:
```diff
@@ -201,9 +202,8 @@ struct SendDataForFluxes {
-  template <typename DbTags, typename... InboxTags,
-            typename ArrayIndex, typename ActionList,
-            typename ParallelComponent>
+  template <typename DbTags, typename... InboxTags, typename ArrayIndex,
```
Why?

nilsdeppe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp:
```diff
@@ -303,10 +303,10 @@ struct SendDataForFluxes {
-                    db::item_type<Tags::VariablesBoundaryData, DbTags>*>
-                    mortar_data) noexcept {
+            [&mortar_id, &temporal_id,
+             &local_data ](const gsl::not_null<
```
Okay, I'm going to stop commenting on this. Stop formatting entire files, only format the things you actually touched.

nilsdeppe reviewed (changes requested):
I stopped after two commits. There are way too many unnecessary changes in this PR. Please stop formatting all the files for no reason. Format the code you touched, and even then during a rename it's better to minimize changes. Once you've removed all the unnecessary changes I'll actually review the rest of it.

nilsvu committed:
Add Mesh tag and interface compute item

nilsvu committed:
Refactor definite_integral to use Mesh

nilsvu committed:
Refactor linearize to use Mesh

nilsvu committed:
Remove Basis::lgl namespace, replaced by Spectral

Instead of calls to Basis::lgl use the Spectral functions now that come
in two versions: One is templated on the Basis and Quadrature, and the
other takes a Mesh<1>. The latter is preferred, so e.g. to retrieve
collocation points in dimension d call
Spectral::collocation_points(mesh.slice_through(d)).

nilsvu committed:
Move Mesh from DataStructures to Domain

nilsvu committed:
Refactor IrregularInterpolant to use Mesh

nilsvu committed:
Test IrregularInterpolant on non-LGL meshes

nilsvu committed:
Refactor logical_coordinates to use Mesh

nilsvu committed:
Refactor partial_derivatives to use Mesh

nilsvu commented:
Done and rebased on `develop`. IWYU complains about Equations.hpp that I didn't touch in this PR (and I think I shouldn't?).

kidder commented:
you touched the corresponding cpp files which triggers IWYU on the hpp files, so please deal with the complaints

nilsdeppe commented:
Those may not be correct, so be sure to understand the suggestions.

nilsdeppe commented:
I'll try to look at this tomorrow morning or evening, sorry for the delay on it!

nilsdeppe commented on src/NumericalAlgorithms/LinearOperators/Linearize.cpp:
```diff
@@ -6,39 +6,41 @@
-      });
-  return apply_matrices(linear_filter_matrices, u, extents);
+DataVector linearize(const DataVector& u, const Mesh<Dim>& mesh) {
+  const Matrix empty{};
```
Why this change?

nilsvu replied:
The extents are not sufficient anymore for the call to get the matrix, instead it needs `mesh.slice_through(d)` for `d=0..Dim`. Do you have a suggestion on how to do this in a more concise implementation? What I wrote is in line with the definition of the function below, and other implementations I considered like using `std::iota` to build an array of dimensions and then mapping over it only appeared more cumbersome.

nilsdeppe replied:
Ah, I see. Nothing obvious comes to mind :)

nilsdeppe commented on tests/Unit/DataStructures/Test_Mesh.cpp:
```diff
@@ -223,6 +225,10 @@ SPECTRE_TEST_CASE("Unit.DataStructures.Mesh", "[DataStructures][Unit]") {
 }
 
+SPECTRE_TEST_CASE("Unit.Domain.MeshTag", "[Unit][Domain]") {
+  CHECK(Tags::Mesh<3>::name() == "Mesh");
```
Please put this in the `Mesh` test so we don't have to spend time waiting for an additional test to start. Tests tend to have a fairly high overhead on some HPC systems.

nilsdeppe commented on tests/Unit/Domain/Test_Element.cpp:
```diff
@@ -65,3 +67,7 @@ SPECTRE_TEST_CASE("Unit.Domain.Element", "[Domain][Unit]") {
 }
+
+SPECTRE_TEST_CASE("Unit.Domain.ElementTag", "[Unit][Domain]") {
+  CHECK(Tags::Element<3>::name() == "Element");
```
again

nilsdeppe commented on tests/Unit/Domain/Test_ElementMap.cpp:
```diff
@@ -190,3 +192,7 @@ SPECTRE_TEST_CASE("Unit.Domain.ElementMap", "[Unit][Domain]") {
 }
+
+SPECTRE_TEST_CASE("Unit.Domain.ElementMapTag", "[Unit][Domain]") {
+  CHECK(Tags::ElementMap<3>::name() == "ElementMap");
```
again

nilsdeppe commented on tests/Unit/Domain/Test_LogicalCoordinates.cpp:
```diff
@@ -321,3 +323,7 @@ SPECTRE_TEST_CASE("Unit.Domain.InterfaceLogicalCoordinates", "[Domain][Unit]") {
 }
+
+SPECTRE_TEST_CASE("Unit.Domain.LogicalCoordinatesTag", "[Unit][Domain]") {
+  CHECK(Tags::LogicalCoordinates<3>::name() == "LogicalCoordinates");
```
again

nilsdeppe reviewed (changes requested):
Couple minor things, feel free to squash and rebase. If you do this before 6am Eastern time I'll look at it in the morning :)

nilsvu committed:
Test names of Domain Tags

nilsdeppe commented:
@nilsleiffischer there appears to IWYU complaints

nilsvu commented:
fixed it @nilsdeppe

kidder commented:
@gsb76 please look at
